contracts/starknet/jug.cairo

```diff
@@ -0,0 +1,433 @@
+@event
+func Deny(usr: felt) {
+}
+// event Init(bytes32 indexed ilk);
+@event
+func Init(ilk: felt) {
+}
+// event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+@event
+func File_duty(ilk: felt, what: felt, data: Uint256) {
+}
+
+// event File(bytes32 indexed what, uint256 data);
+@event
+func File_base(what: felt, data: Uint256) {
+}
+
+// event File(bytes32 indexed what, address data);
+@event
+func File(what: felt, data: felt) {
+}
+
+// event Drip(bytes32 indexed ilk);
+@event
+func Drip(ilk: felt) {
+}
+
+// modifier auth {
+//         require(wards[msg.sender] == 1, "Jug/not-authorized");
+//         _;
+//     }
+func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
+    let (caller) = get_caller_address();
+    let (ward) = _wards.read(caller);
+    with_attr error_message("Jug/not-authorized") {
+        assert ward = 1;
+    }
+    return ();
+}
+
+@view
+func ilks{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(i: felt) -> (ilk: Ilk) {
+    let (res) = _ilks.read(i);
+    return (res,);
+}
+
+@view
+func wards{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(user: felt) -> (
+    res: felt
+) {
+    let (res) = _wards.read(user);
+    return (res,);
+}
+
+@view
+func vat{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: felt) {
+    let (res) = _vat.read();
+    return (res,);
+}
+
+// // --- Init ---
+//     constructor(address vat_) {
+//         wards[msg.sender] = 1;
+//         vat = VatLike(vat_);
+//         emit Rely(msg.sender);
+//     }
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    vat_: felt, ward_: felt
+) {
+    _wards.write(ward_, 1);
+
+    _vat.write(vat_);
+
+    Rely.emit(ward_);
+
+    return ();
+}
+
+// // --- Math ---
+//     function _rpow(uint256 x, uint256 n, uint256 b) internal pure returns (uint256 z) {
+//       assembly {
+//         switch x case 0 {switch n case 0 {z := b} default {z := 0}}
+//         default {
+//           switch mod(n, 2) case 0 { z := b } default { z := x }
+//           let half := div(b, 2)  // for rounding.
+//           for { n := div(n, 2) } n { n := div(n,2) } {
+//             let xx := mul(x, x)
+//             if iszero(eq(div(xx, x), x)) { revert(0,0) }
+//             let xxRound := add(xx, half)
+//             if lt(xxRound, xx) { revert(0,0) }
+//             x := div(xxRound, b)
+//             if mod(n,2) {
+//               let zx := mul(z, x)
+//               if and(iszero(iszero(x)), iszero(eq(div(zx, x), z))) { revert(0,0) }
+//               let zxRound := add(zx, half)
+//               if lt(zxRound, zx) { revert(0,0) }
+//               z := div(zxRound, b)
+//             }
+//           }
+//         }
+//       }
+//     }
+func _rpow{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}(x: Uint256, n: Uint256, base: Uint256) -> (z: Uint256) {
```

Comment:
n is time so it should be a `felt`

Reply:
in order to avoid internal conversions we can keep it as a `uint256` in the signature and convert it outside, seems like the best approach to me